### PR TITLE
Polygon support

### PR DESCRIPTION
**Pull request by will-moore**
The head of this branch was pushed to npm as version 3.1.0.
This includes all changes currently in https://github.com/ome/omero-figure/pull/253 which loads shape-editor 3.1.0 from npm.

### CODE REVIEW

File: src/js/shapeManager.js [hunk 1]
```diff
@@ -83,12 +85,17 @@ var ShapeManager = function ShapeManager(elementId, width, height, options) {
             function(){
                 self.stopDrag.apply(self, arguments);
             });
+        // this.newShapeBg.click(
```
Comment (jburel) [inline src/js/shapeManager.js]: this can be removed

File: src/js/shapes/polygon.js [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+    this.manager = options.manager;
+    this.paper = options.paper;
+
+    if (options.id) {
+        this._id = options.id;
+    } else {
+        this._id = this.manager.getRandomId();
+    }
+    this._points = options.points;
+    // this._rotation = options.rotation || 0;
```
Comment (jburel) [inline src/js/shapes/polygon.js]: this can be removed

File: src/js/shapes/polygon.js [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+    this.element.attr({'fill-opacity': 0.01,
+                        'fill': '#fff',
+                        'cursor': 'pointer'});
+
+    if (this.manager.canEdit) {
+        // Drag handling of element
+        this.element.drag(
+            function(dx, dy) {
+                // DRAG, update location and redraw
+                dx = dx / self._zoomFraction;
```
Comment (jburel) [inline src/js/shapes/polygon.js]: can this be ``0``?

File: src/js/shapes/polygon.js [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+    this.createHandles();
+    // and draw the Polygon
+    this.drawShape();
+};
+
+Polygon.prototype.toJson = function toJson() {
+    var rv = {
+        'type': "Polygon",
+        'points': this._points,
+        // 'rotation': this._rotation,
```
Comment (jburel) [inline src/js/shapes/polygon.js]: this can be removed

File: src/js/shapes/polygon.js [hunk 5]
```diff
@@ -0,0 +1,356 @@
...
+    var strokeColor = this._strokeColor,
+        strokeW = this._strokeWidth * this._zoomFraction;
+
+    var f = this._zoomFraction;
+    var path = this.getPath();
+
+    this.element.attr({'path': path,
+                       'stroke': strokeColor,
+                       'stroke-width': strokeW});
+    // this.element.transform('r'+ this._rotation);
```
Comment (jburel) [inline src/js/shapes/polygon.js]: same

File: src/js/shapes/polygon.js [hunk 6]
```diff
@@ -0,0 +1,356 @@
...
+        hx, hy, handle;
+    this._points.split(" ").forEach(function(xy, i){
+        var xy = xy.split(",");
+        hx = parseInt(xy[0]);
+        hy = parseInt(xy[1]);
+
+        handle = self.paper.rect(hx-hsize/2, hy-hsize/2, hsize, hsize);
+        handle.attr({'cursor': 'move'});
+        handle.h_id = i;
+        // handle.line = self;
```
Comment (jburel) [inline src/js/shapes/polygon.js]: to remove

**Comment by jburel** [comment]
You indicate that shape-editor was pushed to npm.
We will need to formalize that workflows similar to what we did for pip
Are you the sole owner of the account?
since pushing to npm will be in that case part of the release process

**Comment by will-moore** [comment]
Yes - I am the sole owner.

```
$ npm view ome-shape-editor
 
{ name: 'ome-shape-editor',
  'dist-tags': { latest: '3.1.0' },
  versions: [ '3.1.0' ],
  time: 
   { created: '2018-03-01T13:42:29.322Z',
     '3.1.0': '2018-03-01T13:42:29.389Z',
     modified: '2018-03-01T13:42:29.820Z' },
  maintainers: [ 'will-moore <w.moore@dundee.ac.uk>' ],
...
```

If you create an account https://docs.npmjs.com/cli/adduser I can add you as an owner https://docs.npmjs.com/cli/owner

**Comment by jburel** [comment]
now we have a mismatch the version on npm and no tag on github
so we will need to sort it out

**Comment by will-moore** [comment]
As discussed briefly... Any changes to this PR, and merging with develop will make a version 3.1.1 that will need pushing to npm. But, we haven't tagged 3.1.0 on github yet.
Solution is to merge this PR as it is, tag that with 3.1.0.
Then do fixes in another PR, merge and tag with 3.1.1 and push this to npm.
Then we can update OMERO.figure to use that.

**Comment by jburel** [comment]
Merging as it is to match publicly available version

**Pull request merged**